drivers/timer/mcux_lptmr_timer.c

```diff
@@ -32,16 +35,94 @@ BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1,
 #define CYCLES_PER_TICK ((uint32_t)((uint64_t)sys_clock_hw_cycles_per_sec() \
 			/ (uint64_t)CONFIG_SYS_CLOCK_TICKS_PER_SEC))
 
+#define COUNTER_MAX 0xffffffff
+
+#define MAX_TICKS ((COUNTER_MAX / CYCLES_PER_TICK) - 1)
+#define MAX_CYCLES (MAX_TICKS * CYCLES_PER_TICK)
+#define MIN_DELAY  1000
+
+#ifdef CONFIG_MCUX_LPTMR_TIMER_SAFETY_WINDOW_CYCLES
+#define SAFETY_WINDOW_CYCLES CONFIG_MCUX_LPTMR_TIMER_SAFETY_WINDOW_CYCLES
+#else
+#define SAFETY_WINDOW_CYCLES 10
+#endif
+
 /* 32 bit cycle counter */
 static volatile uint32_t cycles;
 
+/*
+ * Stores the current number of cycles the system has had announced to it,
+ * since the last rollover of the free running counter.
+ */
+static uint32_t announced_cycles;
+
+/* Lock on shared variables */
+static struct k_spinlock lock;
+
+static void lptmr_set_safe_immediate(uint32_t target_cycles)
+{
+	uint32_t hw_counter;
+
+	/* Read current hardware counter */
+	hw_counter = LPTMR_GetCurrentTimerCount(LPTMR_BASE);
+
+	/* adjust target to be outside of safety window */
+	if ((target_cycles > hw_counter) &&
+			((target_cycles - hw_counter) <= SAFETY_WINDOW_CYCLES)) {
+		target_cycles = hw_counter + SAFETY_WINDOW_CYCLES + 1;
+	}
+
+	LPTMR_SetTimerPeriod(LPTMR_BASE, target_cycles);
+}
+
 void sys_clock_set_timeout(int32_t ticks, bool idle)
 {
 	ARG_UNUSED(idle);
 
 	if (idle && (ticks == K_TICKS_FOREVER)) {
 		LPTMR_DisableInterrupts(LPTMR_BASE, kLPTMR_TimerInterruptEnable);
 	}
+
+	if (!IS_ENABLED(CONFIG_TICKLESS_KERNEL)) {
+		return;
+	}
+
+	k_spinlock_key_t key;
+	uint32_t next, adj, now;
+
+	ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : ticks;
+	/* Clamp ticks. We subtract one since we round up to next tick */
+	ticks = CLAMP((ticks - 1), 0, (int32_t)MAX_TICKS);
+
+	key = k_spin_lock(&lock);
+
+	/* Read current timer value */
+	now = LPTMR_GetCurrentTimerCount(LPTMR_BASE);
+
+	/* Adjustment value, used to ensure next capture is on tick boundary */
+	adj = (now - announced_cycles) + (CYCLES_PER_TICK - 1);
+
+	next = ticks * CYCLES_PER_TICK;
+	/*
+	 * The following section rounds the capture value up to the next tick
+	 * boundary
+	 */
+	if (next <= MAX_CYCLES - adj) {
+		next += adj;
+	} else {
+		next = MAX_CYCLES;
+	}
+	next = (next / CYCLES_PER_TICK) * CYCLES_PER_TICK;
+
+	if ((int32_t)(next + announced_cycles - now) < MIN_DELAY) {
+		next += CYCLES_PER_TICK;
+	}
+
+	next += announced_cycles;
```

Comment:
could this `next` overflow? which means in function `lptmr_set_safe_immediate`, target_cycles < hw_counter